app/controllers/application_controller.rb

```diff
@@ -221,8 +226,13 @@ def store_location
   # Redirect to the URI stored by the most recent store_location call or
   # to the passed default.
   def redirect_back_or_default(default = root_path)
+    back_url = params[:user_session][:return_to_url] if params[:user_session]
```

Comment:
This is an unsafe redirect. Because it's in the params, it can be used for phishing (because you can send someone a link that looks like AO3 but will secretly redirect them elsewhere) or other malicious activities. When a URL is passed through params like this, it needs to be checked before it's passed to `redirect_to`.

(This also has the same issue as [AO3-5949](https://otwarchive.atlassian.net/browse/AO3-5949): there are [some](https://otwarchive.atlassian.net/browse/AO3-5953) [actions](https://otwarchive.atlassian.net/browse/AO3-5685) that really shouldn't be GETs, and I'm not sure if it's a good idea to use params to redirect before that's resolved.)

Reply:
Would it be suffficant to check that the url is relative ( I mean with no host portion ? , or with the host portion set to nil  ? )

Reply:
Host, scheme, maybe also port/username/password? It'd probably be better to pick a few whitelisted components that are allowed to have values, and reject any URLs that have any other components set.

But as I said, this has [AO3-5953](https://otwarchive.atlassian.net/browse/AO3-5953) and [AO3-5685](https://otwarchive.atlassian.net/browse/AO3-5685) as pre-requisites, so there's no rush to fix it.